feat: add enriched error interface to flux

jsternberg opened the pull request:
We need a common way of categorizing and understanding errors returned
from InfluxDB. We model the errors based on the same concepts that
InfluxDB uses, but we create our own error codes that are modeled on the
grpc error codes.

Each native flux error will have a code that represents the category or
type of error and a message that gives detailed and human-readable
output about the error.

The errors library is included as an internal library and the error
struct itself is exposed using a type alias in the top level package.
This allows the errors to be used by packages that cannot depend on the
top level package (like semantic), but still have the error type be
exposed.

### Done checklist
- [x] docs/SPEC.md updated
- [x] Test cases written

Related to #1364.

nathanielc commented:
It seems there are two error mechanisms in InfluxDB, the one you have copied here and this one  https://github.com/influxdata/influxdb/tree/master/kit/errors 

What is the difference?

nathanielc commented on errors.go:
```diff
@@ -0,0 +1,244 @@
+	EForbidden           = "forbidden"
+	ETooManyRequests     = "too many requests"
+	EUnauthorized        = "unauthorized"
+	EMethodNotAllowed    = "method not allowed"
```
These error code seem very specific to HTTP and that does not apply well to Flux as the repo itself does not interact with HTTP APIs.

jsternberg replied:
I agree on some of them. `EMethodNotAllowed` is a bit _too_ http specific. Forbidden and unauthorized are decently generic even if they map directly to HTTP. Too many requests is technically generic even if it is directly mapped to http.

One possible issue though is, a little by necessity, these need to be the same as influxdb. I guess we can remove the ones that aren't used or are more egregious like method not allowed, but I think that's a judgment call.

nathanielc replied:
Why do we need to the same codes as influxdb? That is not clear to me.

Also it looks like the methods below have been implemented by using type assertions on anonymous interfaces so copying the types should also be safe. Not translation should be needed from errors originating in Flux vs errors originating in Influxdb.

jsternberg replied:
It's because of serialization. If we don't have the same codes as influxdb, then they won't be able to read our error codes properly.

They'll do something like:

```go
if influxdb.ErrorCode(err) == influxdb.EForbidden {
    // ...
}
```

And even if we returned `EForbidden`, it won't recognize it. Additionally, JSON serialization, which is very possible that this will go through, will cause a `flux.Error` to get reconstructed as an `influxdb.Error` so it wouldn't be possible to differentiate the two by what type they are. We cannot make the assumption a `flux.Error` will stay a `flux.Error`.

nathanielc replied:
I see, I had assumed that these error were meant to be opaque to the code, but looking through the influxdb code I see that is not the case. I think we should add a comment explicitly stating that since common Go error handling is that errors are opaque.

Is there a different approach we can take that doesn't tightly couple error handling in Flux with how InfluxDB does error handling?


As an aside:

> We cannot make the assumption a flux.Error will stay a flux.Error.

This is a good thing in my mind. It means the errors are defined by their behavior and not their representation.

jsternberg replied:
I think it would be best to take the core of the errors library from influxdb and put it into its own separate and versioned package. The utility methods that influxdb has don't have to come along with it for simplicity and to ensure the interface doesn't change. That way we have no dependency on influxdb and we can type alias all of the functions we want or any new ones.

The error codes are supposed to be generic similar to HTTP (although the codes are not supposed to be HTTP) so they aren't supposed to be influxdb specific. There's a note about adding them to the swagger doc, but that could be changed to a go generate command that reads a file from a dependency in influxdb.

jsternberg replied:
For clarity, I don't think using the same error handling would tightly couple the two. Once it is its own library, it stands on its own and it just happens to be that the two use the same method of reporting errors.

nathanielc replied:
I see two ways this could work:

1. Couple the error handling in which case it makes sense to create a separate package that flux and influxdb both depend on. Otherwise the divergence would create a nightmare of bugs.
2. Decouple the error handling, meaning do not use the same error codes or types, but just take the same basic strategy. Flux can export its error codes and InfluxDB can make an explicit translation between them if needed. I would rather Flux errors be opaque to InfluxDB with a few explicit exceptions. This way packages consuming Flux (i.e. InfluxDB) can make decisions about what Flux error codes mean and if they will take action on them, otherwise they can treat them as opaque. The only current use case will be for Flux to report errors as either system errors or user errors, which InfluxDB will inspect to update appropriate statistics.

I prefer option 2 as it reduces the surface area of the Flux Go API. 

I have a few concerns over how option 1 could play out in confusing and negative ways. 

* Since InfluxDB and Flux will use the same error codes, the set of error codes will be the union of the two needs. In addition to the error code being part of the  InfluxDB public API this means that changes need to be strictly controlled and may not be possible to change in some cases. These two combined leads to a tight coupling that makes it hard to make changes to either project.
* All consumers of Flux as a library will need to make a decision about how to handle the error codes Flux can produce. As such that set of error codes should be as small as possible, including influxdb specific codes is confusing and error prone to consumers of Flux.
* This tight coupling can cause very hard to debug and unexpected issues. For example given that InfluxDB has logic that conditions on the error codes of various errors. And given that Flux produces errors with error codes that InfluxDB consumes. A change to errors either in Flux code or anywhere in the call stack between Flux and InfluxDB can trigger different logic in InfluxDB. These kinds of changes will have very indirect and hard to predict behaviors as the logic being affected is far away from the changes being made. 

Said another way, in general the practice of inspecting errors creates the possibility that changes can effect code that is loosely related (i.e. way up the call stack). As such I want to tightly control that surface area so it is more clear how said changes effect the code base as a whole.

jsternberg replied:
Ok. I think this means the interfaces aren't really needed then because we would be saying the API is completely different so there's no expectation the values are the same. I think that means we'll need to add a method to influxdb so if it sees a `*flux.Error` it will translate it to an `*influxdb.Error`. I took another look at the source blog post for our system ([here](https://middlemost.com/failure-is-your-domain/) and that's what it said to do so I think that's a good solution.

goller replied:
> Since InfluxDB and Flux will use the same error codes, the set of error codes will be the union of the two needs. In addition to the error code being part of the InfluxDB public API this means that changes need to be strictly controlled and may not be possible to change in some cases. These two combined leads to a tight coupling that makes it hard to make changes to either project.

I don't think this will be a problem. Error codes are for groups of problems that inform consumers of the library what to do. Except the `method not available` I think each of the error codes in errors.go today are good groupings and each are applicable for flux. 

I think method not available should be removed and I'll make a PR to do so.

What groupings of errors do you imagine are not covered?  Flux is integral to influxdb and if influxdb does not support an error group flux needs, then, it absolutely should be added.

>    All consumers of Flux as a library will need to make a decision about how to handle the error codes Flux can produce. As such that set of error codes should be as small as possible, including influxdb specific codes is confusing and error prone to consumers of Flux.

I don't really see any error groupings that would not be appropriate for flux except method not allowed.

>    This tight coupling can cause very hard to debug and unexpected issues. For example given that InfluxDB has logic that conditions on the error codes of various errors. And given that Flux produces errors with error codes that InfluxDB consumes. A change to errors either in Flux code or anywhere in the call stack between Flux and InfluxDB can trigger different logic in InfluxDB. These kinds of changes will have very indirect and hard to predict behaviors as the logic being affected is far away from the changes being made.

Ultimately, the behavior that would change in influxdb from error codes would be the HTTP status code from /api/v2/query.  I think it is quite reasonable for flux to govern /api/v2/query

jsternberg commented:
@nathanielc the error library in `kit/errors` appears to be a first draft and is completely unused beyond accidental usage from `goimports`. I'll confirm that with @goller and then remove it from the repository.

jsternberg commented:
I have removed the extraneous error codes and removed the interfaces so this remains its own distinct error type and we will translate this error to an `*influxdb.Error` inside of influxdb.

I have removed all but three of the error types. I don't think this will be enough, but I'm not going to try and get the correct level of what we need versus generic error codes in this pull request.

jsternberg commented:
We're going to discuss this in a meeting sometime next week to hash out the details and also how it integrates with go 1.13's handling of errors.

jsternberg commented:
@nathanielc I copied over the error codes from grpc. I filtered out the ones I didn't think made sense inside of flux and modified the style to match how we display error messages. I removed op and the extra code that dealt with recursive behavior and instead opted to just implement `Unwrap`. I removed the json serialization, but I plan to add that back.

nathanielc commented on codes/codes.go:
```diff
@@ -0,0 +1,172 @@
+	Unknown Code = 2
+
+	// InvalidArgument indicates client specified an invalid argument.
+	// Note that this differs from FailedPrecondition. It indicates arguments
```
Several of these comments refer to `FailedPrecondition` but that is not one of the errors brought over. Can you update the comments or add `FailedPrecondition` back in?

jsternberg replied:
I'm going to look at `FailedPrecondition` again. It seems to be mentioned a lot and I didn't understand it at first. I'll update these comments regardless, but I want a better reason why I think `FailedPrecondition` shouldn't be there other than, "I didn't understand what it meant and it didn't seem to apply to any situations I could think of."

nathanielc commented on codes/codes.go:
```diff
@@ -0,0 +1,172 @@
+	//
+	// See litmus test above for deciding between FailedPrecondition,
+	// Aborted, and Unavailable.
+	Unavailable Code = 12
```
I really like the distinction between Internal and Unavailable. It allows us to communicate that one of our dependencies is failing but otherwise our system is working as expected.

nathanielc commented on errors.go:
```diff
@@ -0,0 +1,34 @@
+// Error is the error struct of flux.
+type Error struct {
+	Code codes.Code
+	Msg  string
```
Should we add a comment that we expect Code and Msg to always be set and that Err is optional and only needed to wrap an existing error?

Also we should state that only the `Code` field should be inspected by calling code.

jsternberg replied:
Yea, I think so. Although message is also going to be optional. There may be situations where we can't add any additional context.

jsternberg commented:
I was trying this out on Wednesday to see how it would be used and I found that the errors don't seem to cover the generic categories nearly as well as I thought. In particular, I found that including the name of the error in the error text was mostly ineffective and created bad error messages. Here was how a type error code changed:

```
invalid argument: type error 3:1-3:23: invalid record access "_value": int != float
```

The `invalid argument` seemed the best choice based on our conversations, but the error message seemed like complete nonsense to me. I think it might be best not to include the code in the error message unless there is no message.

jsternberg commented on codes/codes.go:
```diff
@@ -0,0 +1,229 @@
+	// may be converted to this error.
+	Unknown Code = 2
+
+	// InvalidArgument indicates client specified an invalid argument.
```
This name has made no sense while I have been playing around with reporting errors. I'm renaming it to `Invalid` to indicate that the client gave us something that was not valid rather than the specific "invalid argument" which doesn't make sense in contexts where there is no argument or the argument is missing.

jsternberg commented:
I added tests for all of the errors stuff and also added a go 1.13 specific test which passes when using go 1.13, but doesn't run with go 1.12.
